README
This is a plugin for Pidgin [1] that allows you to display LaTeX [2] output into your IMs.

Important : this plugin is not compatible with gaim-encryption plugin. 
If you have this plugin, please be sure not having the 2 working at the same time. 

You need to have mimetex (1.64 preferred,1.50 is definitely to old) [3] or mathtex [4].

WINDOWS:
Only mimetex is supported in windows. Copy the mimetex.bat file somewhere within your path
(Placing it in C:\Windows\ would work) and edit the path to mimetex.exe depending 
on where you have placed it. Mimetex may *not* be in C:\Windows. It is recommended that
you place it in the plugin folder.
A compiled binary of mimetex is available on it's official homepage [3].

How to choose:
1. mimetex, + Small, no extra dependencies.
   mathtex, - Requires latex and dvipng.

2. mimetex, + Very fast.
   mathtex,   Takes 1-2 seconds first image, then quite fast.

3. mimetex, - At best only produces gifs anti aliased to white or the inversed font color.
              (but with one color transparency).
   mathtex, + generates PNG-images with true transparency.

4. mimetex, - Not 100% compatible with LaTeX.
   mathtex, + 100% compatible.

In addition, mathtex renders nicer looking graphics
If you don't have latex installed, you probably want mimetex, if you
have latex, then mathtex and dvipng.

LINUX:
To build and install :
You can compile the plugin using
$ make
and install it with
$ make install
This will install it in ~/.purple/plugins so that only the user who install it can use it.

To install it for everybody on your computer,
$ make
$ su
# make install PREFIX="/path/to/pidgin" (this command as root user)
generally /path/to/pidgin is /usr or /usr/local. 
If you don't know the path then you can find out using
$ whereis pidgin
and look for the part before "/lib/pidgin".

MINGW:
-- HELP WANTED ---

To use
- put LaTeX code into $$ ... $$ markup, for example $$x \in \mathbb{R}$$

[1] http://www.pidgin.im/
[2] http://www.latex-project.org/
[3] http://www.forkosh.com/mimetex.html
[4] http://www.forkosh.com/mathtex.html

Developer : Mikael Öhman <micketeer@gmail.com>


CHANGELOG
= Details =
Version 1.0.6
  * Change " to ' when sending system calls. Stops the terminal from removing an escapesequence, namely  \\ to \. This should make it easier to have linebreaks in for example arrays.

Version 1.0.5
  * Fixed Ubuntu-bug where system() returned errors. There are still errors, but i'm just ignoring them at they still seem to work

Version 1.0.4
  * Fixed makefile to only require purple-dev, not pidgin-dev
  * Support for windows. See wiki pages or README

Version 1.0.3
  * Fixed cross compilation for Win32
  * Fixed escaped apostrohpes (prime sign ')
  * Added DEBUG option in makefile. Prints out debug information if compiled with make DEBUG=1

Version 1.0.2
  * Fixed bug where incoming message were not rendered.
  * Improved error message for missing renderer.
  * Fixed silly mistake in header file
  * Changed homepage

Version 1.0.1
  * Warning message when image can't be sent.
  * Redesign of signal handling.
  * Removed potential (but inlikely) memory leak.

Version 1.0.0
  * Initial version
  * Made pidgin-mimetex from pidgin-LaTeX


pidginTeX.h
/* pidgin-mimetex.h 
 * Based on:
 * LaTeX.c
 * pidgin-latex plugin
 *
 * This a plugin for Pidgin to display LaTeX formula in conversation
 * PLEASE, send any comment, bug report, etc. to the trackers at sourceforge.net
 * Copyright (C) 2006-2007 Benjamin Moll (qjuh@users.sourceforge.net)
 * some portions : Copyright (C) 2004-2006 Nicolas Schoonbroodt (nicolas@ffsa.be)
 *                 Copyright (C) 2004-2006 GRIm@ (thegrima@altern.org).
 * 		   Copyright (C) 2004-2006 Eric Betts (bettse@onid.orst.edu).
 * Windows port  : Copyright (C) 2005-2006 Nicolai Stange (nic-stange@t-online.de)
 * Other portions heavily inspired and copied from gaim sources
 * Copyright (C) 1998-2007 Pidgin developers pidgin.im
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; This document is under the scope of
 * the version 2 of the License, or any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program (see COPYING); if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#ifndef _PIDGINTEX_H_
#define _PIDGINTEX_H_
#ifndef G_GNUC_NULL_TERMINATED
#  if __GNUC__ >= 4
#    define G_GNUC_NULL_TERMINATED __attribute__((__sentinel__))
#  else
#    define G_GNUC_NULL_TERMINATED
#  endif /* __GNUC__ >= 4 */
#endif /* G_GNUC_NULL_TERMINATED */

#define PURPLE_PLUGINS

#ifdef ENABLE_NLS
#  include <locale.h>
#  include <libintl.h>
#  define _(String) dgettext (PLUGIN_NAME, String)
// TODO:
#  define LOCALEDIR "locale"
#else
#  define _(String) (String)
#endif

// Includes
#ifdef _WIN32
#define g_file_get_contents g_file_get_contents_utf8
#include <windows.h>
#endif

#include <conversation.h>
#include <debug.h>
#include <signals.h>
#include <imgstore.h>
#include <util.h>
#include <notify.h>
#include <server.h>
#include <log.h>
#include <version.h>

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>

// Constant
#define TEX_DELIMITER    "$$"
#define PREFS_BASE       "/plugins/core/" PLUGIN_NAME
#define PREFS_SENDIMAGE  (PREFS_BASE "/sendimage")
#define PREFS_PREPEND    (PREFS_BASE "/prepend")
#define PREFS_RENDERER   (PREFS_BASE "/renderer")
#define PREFS_PRINTEXPR  (PREFS_BASE "/printexpr")
#define PREFS_NEGATE     (PREFS_BASE "/negate")
#define PREFS_SMASH      (PREFS_BASE "/smash")
#define PREFS_STYLE      (PREFS_BASE "/style")
#define PREFS_FONT_SIZE  (PREFS_BASE "/fontsize")
#define PREFS_FONT_COLOR (PREFS_BASE "/fontcolor")

/* font sizes for mathTeX */
static char* mathfont[] = {"tiny","footnotesize","normalsize","large",
                 "Large","LARGE","huge","Huge"};

/* 
 * latex_to_image creates an image with the LaTeX code pointed by *latex
 * *latex points to the latex string
 * **file_img receives filename of the temporary generated image; 
 * file must be deleted and string must be freed by caller
 *
 * returns TRUE on success, false otherwise
 */
//static gboolean latex_to_image(char *latex, char **file_img);

/*
 * Transform *tmp2 extracting some *startdelim here *enddelim thing,
 * make an image from latex-input and tmp2 becomes 'some <img="number"> thing'
 *
 * returns TRUE on success, FALSE otherwise
 */
//static gboolean analyse(char **msg, char *startdelim, char *enddelim);

/*
 * log_write performs the effective write of the latex code in the IM windows
 * 	*conv is a pointer onto the conversation context
 *	*nom is the name of the correspondent
 *	*message is the modified message with the image
 *	*messFlags is Flags related to the messages
 *	*original is the original message unmodified
 *
 * returns TRUE.
 */
//static void pidgin_latex_write(PurpleConversation *conv, const char *nom, 
//    char *message, PurpleMessageFlags messFlag, char *original);

/* to intercept in- and outgoing messages */
static gboolean message_write(PurpleAccount *account, const char *sender, char **message, 
    PurpleConversation *conv, PurpleMessageFlags flags);
static void message_wrote(PurpleAccount *account, const char *sender, 
    const char *message, PurpleConversation *conv, PurpleMessageFlags flags);
static void message_send(PurpleAccount *account, char *recipient, char **message);

/* variables for callback functions */
static gboolean logflag;
static char *originalmsg;
static char *modifiedmsg;

/*
 * searchPATH searches the PATH-environment for the specified file file.
 * *file is the name of the executable to search for
 *
 * returns the right full path, e.g. with the executable's name appended, 
 * NULL on failure, must be freed with free()
 */
char* searchPATH(const char *file);

/*
 * execute executes the *cmd.
 * Advantage to system()-call under win32 ist that no console window pops up.
 * On systems other than windows this function is just a wrapper around the system()-call.
 * *cmd is the command to execute.
 *
 * returns -1 if execution failed, otherwise the return code of the executed program.
*/
static int execute(char *cmd);

/*
 * win32_purple_notify extracts error information of last occured 
 * WIN32-Error due to an API-call and asserts it to gaim via purple_notify_error
 * Error message will be prepended by *prep:
 */
//#ifdef _WIN32
//void win32_purple_notify_error(char *prep);
//#endif

#endif


pidginTeX.c
/* pidginTeX.c
 * Mikael Öhman (micketeer@gmail.com)
 * Built from pidgin-latex plugin
 * PLEASE, send any comment, bug report, etc. to the trackers at googlecode
 *
 * original pidgin-latex plugin writers:
 * Copyright (C) 2006-2007 Benjamin Moll (qjuh@users.sourceforge.net)
 * some portions : Copyright (C) 2004-2006 Nicolas Schoonbroodt (nicolas@ffsa.be)
 *                 Copyright (C) 2004-2006 GRIm@ (thegrima@altern.org).
 *         Copyright (C) 2004-2006 Eric Betts (bettse@onid.orst.edu).
 * Windows port  : Copyright (C) 2005-2006 Nicolai Stange (nic-stange@t-online.de)
 * Other portions heavily inspired and copied from gaim sources
 * Copyright (C) 1998-2007 Pidgin developers pidgin.im
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; This document is under the scope of
 * the version 2 of the License, or any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program (see COPYING); if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "pidginTeX.h"
#include <errno.h>

static void open_log(PurpleConversation *conv)
{
    conv->logs = g_list_append(NULL, purple_log_new(conv->type == PURPLE_CONV_TYPE_CHAT ? 
                PURPLE_LOG_CHAT : PURPLE_LOG_IM, 
                conv->name, conv->account, conv, time(NULL), NULL));
}

char* searchPATH(const char *file)
{
    char* searchexpr = g_strdup_printf("which %s > /dev/null",file);
    int rt = system(searchexpr);
    if (rt == -1)
        perror("Critical error: ");
    else
        WEXITSTATUS(rt);
    DEBUG_PRINT(stderr, "Performed command:\n%s\nreturn value was %d\n",searchexpr,rt);
    free(searchexpr);
    // Ubuntu fix!
    // This is a horrible fix for Ubuntu, where system appearantly fails now and then. 
    // I have not been able to find a reason for this, but the commands seem to run anyway
    return rt && rt != -1 ? NULL : g_strdup(file);
}

static int execute(char *cmd)
{
#ifdef _WIN32
    DWORD exitcode = 0;
    STARTUPINFO sup;
    PROCESS_INFORMATION pi;
    ZeroMemory( &sup, sizeof(sup) );
    ZeroMemory( &pi,  sizeof(pi) );
    sup.cb          = sizeof(sup);
    sup.wShowWindow = SW_HIDE;
    sup.dwFlags     = STARTF_USESHOWWINDOW;

    if (!CreateProcess(NULL, cmd, NULL, NULL, TRUE, 0, NULL, NULL, &sup, &pi))
    {
        purple_notify_error(NULL, PLUGIN_NAME, cmd, NULL);
        return -1;
    }
    if (WAIT_OBJECT_0!=WaitForSingleObjectEx(pi.hProcess, INFINITE, FALSE))
    {
        purple_notify_error(NULL, PLUGIN_NAME, cmd, NULL);
        CloseHandle(pi.hProcess);
        CloseHandle(pi.hThread);
        return -1;
    }
    do
    {
        if (!GetExitCodeThread(pi.hThread, &exitcode))
        {
            purple_notify_error(NULL, PLUGIN_NAME, cmd, NULL);
            return -1;
        }
        Sleep(10);
    } while(exitcode==STILL_ACTIVE);
    CloseHandle(pi.hProcess);
    CloseHandle(pi.hThread);
    return exitcode;
#else
    return system(cmd);
#endif
}

static gboolean latex_to_image(char *tex, char **file_img)
{
    FILE* tmpfile = purple_mkstemp(file_img,TRUE);
    if(!*file_img)
    {
        purple_notify_error(NULL, PLUGIN_NAME, _("Couldn't create temporary file"), NULL);
        unlink(*file_img);
        free(*file_img);
        return FALSE;
    }
    fclose(tmpfile);

#ifndef _WIN32
    const char *renderer = purple_prefs_get_string(PREFS_RENDERER);
    char *cmdTeX = searchPATH(renderer);
    if (!cmdTeX)
    {
        char *err_msg = !strcmp(renderer,"mimetex") ? 
            _("Failed to find: mimetex\n"
            "Make sure you have it installed or change renderer.") :
            _("Failed to find: mathtex\n"
            "Make sure you have it installed or change renderer.");
        // This didn't work for no good reason.
        //g_strdup_printf("Failed to execute command: %s."
        //" Make sure you have it installed or change renderer.",cmdTeX); 
        purple_notify_error(NULL, PLUGIN_NAME, err_msg, NULL);
        free(cmdTeX);
        return FALSE;
    }
#endif

    // Fixes the unusual escaped apostrophe (rest can be handled by mathtex)
    //tex = unescape_apos(tex);
    GRegex* regex = g_regex_new("&apos;", 0, 0, NULL);
    tex = g_regex_replace(regex, tex, -1, 0, "'", 0, NULL);
    g_regex_unref(regex);

    // Build the options the options
    int  fontsize         = purple_prefs_get_int(PREFS_FONT_SIZE);
    const char *prepend   = purple_prefs_get_string(PREFS_PREPEND);
    const char *fontcolor = purple_prefs_get_string(PREFS_FONT_COLOR);
    const char *style     = purple_prefs_get_string(PREFS_STYLE);
    const char *reverse   = purple_prefs_get_bool(PREFS_NEGATE) ? "\\reverse" : "";
    const char *smash     = purple_prefs_get_string(PREFS_SMASH);
    char* cmdparam;

    int usecolor = strlen(fontcolor);
#ifdef _WIN32
    cmdparam = g_strdup_printf("cmd.exe /C mimetex -d -s %d \"%s%s%s%s%s{%s %s}\" %s",
        fontsize, usecolor ? "\\":"", usecolor ? fontcolor:"",
        reverse, style, smash, prepend, tex, *file_img);
#else
    if (!strcmp(renderer, "mimetex"))
    {
        cmdparam = g_strdup_printf("\"%s\" -d -s %d '%s%s%s%s%s{%s %s}' > %s",
            cmdTeX, fontsize, usecolor ? "\\":"", usecolor ? fontcolor:"", 
            reverse, style, smash, prepend, tex, *file_img);
    }
    else //if (!strcmp(renderer,"mathtex"))
    {
        cmdparam = g_strdup_printf( 
            "\"%s\" -m 0 '\\png\\usepackage{color}\\color{%s}%s\\%s %s %s' -o %s",
            cmdTeX, usecolor ? fontcolor : "black", style,
            mathfont[fontsize], prepend, tex, *file_img);
    }
    free(cmdTeX);
#endif
    free(tex);

    DEBUG_PRINT(stderr, "%s\n",cmdparam);
    int rt = execute(cmdparam);
    if (rt == -1 && errno == ECHILD) // See Ubuntu at searchPATH
        DEBUG_PRINT(stderr, "Hoping it works anyway, continuing\n");
    else if(rt)
    {
#ifdef _WIN32
        purple_notify_error(NULL, PLUGIN_NAME, _("Failed to execute renderer,"
            " something might be wrong in the latex expression"), NULL);
#else
        char *err_msg = !strcmp(renderer,"mimetex") ? 
            _("Failed to execute: mimetex\n"
            "Something might be wrong in the latex expression.") :
            _("Failed to execute: mathtex\n"
            "Something might be wrong in the latex expression.");
            // This didn't work for no good reason.
            //g_strdup_printf("Failed to execute command: %s."
            //" Make sure you have it installed or change renderer.",cmdTeX); 
        purple_notify_error(NULL, PLUGIN_NAME, err_msg, NULL);
#endif
        free(cmdparam);
        //free(err_msg);
        return FALSE;
    }
    free(cmdparam);

#ifndef _WIN32
    // Ugly hack becuase mathtex makes up it's own filename (appends fileext). sigh
    if (!strcmp(renderer,"mathtex"))
    {
        char* file_img2 = g_strdup_printf("%s.png",*file_img);
        unlink(*file_img);
        free(*file_img);
        *file_img = file_img2;
    }
#endif
    return TRUE;
}

static gboolean analyse(char **msg, char *startdelim, char *enddelim)
{
    int imgcounter = 0;
    char *ptr1, *ptr2 = NULL;
    while(ptr1 = strstr(*msg, startdelim))
    {
        char *tex, *file_img;
        gchar *filedata;
        size_t size;
        GError *error = NULL;

        int msglen = strlen(*msg);
        tex = &ptr1[strlen(startdelim)];
        if (!(ptr2 = strstr(tex, enddelim))) return FALSE;
        *ptr2 = '\0';

        if (!latex_to_image(tex, &file_img)) return FALSE;

        if (!g_file_get_contents(file_img, &filedata, &size, &error))
        {
            purple_notify_error(NULL, PLUGIN_NAME, error->message, NULL);
            g_error_free(error);
            *ptr2 = *startdelim;
            return FALSE;
        }
        unlink(file_img);
        free(file_img);

        char* name = g_strdup_printf("pidginTeX-%s-%d.png",purple_date_format_long(NULL),
            imgcounter++);
        int idimg = purple_imgstore_add_with_id(filedata, size, name);
        free(name);
        if (idimg == 0)
        {
            purple_notify_error(NULL, PLUGIN_NAME, _("Failed to store image."), NULL);
            *ptr2 = *startdelim;
            return FALSE;
        }

        *ptr1 = '\0';
        char* tmpmsg;
        tmpmsg = g_strdup_printf("%s<img id=\"%d\" alt=\"%s\"> %s%s", *msg,idimg,tex,
                purple_prefs_get_bool(PREFS_PRINTEXPR) ? tex : "", &ptr2[strlen(enddelim)]);
        free(*msg);
        *msg = tmpmsg;
    }
    return TRUE;
}

static gboolean message_write(PurpleAccount *account, const char *sender, 
    char **message, PurpleConversation *conv, PurpleMessageFlags flags)
{
    DEBUG_PRINT(stderr, "message_write = \n%s\n",*message);
    if (!modifiedmsg && strstr(*message,TEX_DELIMITER) && (modifiedmsg = strdup(*message)) &&
        !analyse(&modifiedmsg, TEX_DELIMITER, TEX_DELIMITER))
    {
        free(modifiedmsg);
        modifiedmsg = NULL;
    }
    /*if (!modifiedmsg && strstr(*message,TEX_DELIMITER))
    {
        modifiedmsg = strdup(*message);
        analyse(&modifiedmsg, TEX_DELIMITER, TEX_DELIMITER);
    }*/
    if (modifiedmsg)
    {
        logflag = purple_conversation_is_logging(conv);
        purple_conversation_set_logging(conv, FALSE);
        originalmsg = *message;
        *message    = modifiedmsg;
        modifiedmsg = NULL;
        if (conv->features & PURPLE_CONNECTION_NO_IMAGES && purple_prefs_get_bool(PREFS_SENDIMAGE))
            purple_notify_warning(NULL, PLUGIN_NAME, 
                _("Image was NOT sent, this conversation does not support images."), NULL);
    }
    return FALSE;
}

static void message_wrote(PurpleAccount *account, const char *sender, 
    const char *message, PurpleConversation *conv, PurpleMessageFlags flags)
{
    DEBUG_PRINT(stderr, "message_wrote = \n%s\n",message);
    if (originalmsg && logflag)
    {
        purple_conversation_set_logging(conv, logflag);
        if (!conv->logs)
            open_log(conv);
        GList *log = conv->logs;
        while (log)
        {
            purple_log_write((PurpleLog *)log->data, flags, sender, time(NULL), originalmsg);
            log = log->next;
        }
        free(originalmsg);
        originalmsg = NULL;
    }
}

static void message_send(PurpleAccount *account, char *recipient, char **message)
{
    DEBUG_PRINT(stderr, "message_send = \n%s\n",*message);
    if (!purple_prefs_get_bool(PREFS_SENDIMAGE) || !strstr(*message, TEX_DELIMITER) ||
        !(modifiedmsg = strdup(*message)))
        return;
    else if (analyse(&modifiedmsg, TEX_DELIMITER, TEX_DELIMITER))
        *message = strdup(modifiedmsg);
    else
    {
        free(modifiedmsg);
        modifiedmsg = NULL;
    }
}

static gboolean plugin_load(PurplePlugin *plugin)
{
    void *conv_handle = purple_conversations_get_handle();
    purple_signal_connect(conv_handle,"writing-im-msg",  plugin,PURPLE_CALLBACK(message_write), NULL);
    purple_signal_connect(conv_handle,"writing-chat-msg",plugin,PURPLE_CALLBACK(message_write), NULL);
    purple_signal_connect(conv_handle,"wrote-im-msg",    plugin,PURPLE_CALLBACK(message_wrote), NULL);
    purple_signal_connect(conv_handle,"wrote-chat-msg",  plugin,PURPLE_CALLBACK(message_wrote), NULL);
    purple_signal_connect(conv_handle,"sending-im-msg",  plugin,PURPLE_CALLBACK(message_send), NULL);
    purple_signal_connect(conv_handle,"sending-chat-msg",plugin,PURPLE_CALLBACK(message_send), NULL);
    purple_debug(PURPLE_DEBUG_INFO, PLUGIN_NAME, PLUGIN_NAME " loaded\n");
    modifiedmsg = NULL;
    return TRUE;
}

static gboolean plugin_unload(PurplePlugin * plugin)
{
    void *conv_handle = purple_conversations_get_handle();
    purple_signal_disconnect(conv_handle,"writing-im-msg",  plugin,PURPLE_CALLBACK(message_write));
    purple_signal_disconnect(conv_handle,"writing-chat-mg", plugin,PURPLE_CALLBACK(message_write));
    purple_signal_disconnect(conv_handle,"wrote-im-msg",    plugin,PURPLE_CALLBACK(message_wrote));
    purple_signal_disconnect(conv_handle,"wrote-chat-msg",  plugin,PURPLE_CALLBACK(message_wrote));
    purple_signal_disconnect(conv_handle,"sending-im-msg",  plugin,PURPLE_CALLBACK(message_send));
    purple_signal_disconnect(conv_handle,"sending-chat-msg",plugin,PURPLE_CALLBACK(message_send));
    return TRUE;
}

static PurplePluginPrefFrame * get_plugin_pref_frame(PurplePlugin *plugin)
{
    PurplePluginPrefFrame *frame = purple_plugin_pref_frame_new();
    PurplePluginPref *ppref;

    purple_plugin_pref_frame_add(frame,purple_plugin_pref_new_with_label(
                _("All of these choices will only change your image, not your message.")));

    // Do send image
    ppref = purple_plugin_pref_new_with_name_and_label(
            PREFS_SENDIMAGE,
            _("Send image. Make sure images can be sent before selecting!"));
    purple_plugin_pref_frame_add(frame, ppref);

#ifndef _WIN32
    // Select renderer
    ppref = purple_plugin_pref_new_with_name_and_label(
            PREFS_RENDERER,
            _("Choose a renderer. Program must be installed in path."));
    purple_plugin_pref_set_type(ppref, PURPLE_PLUGIN_PREF_CHOICE);
    purple_plugin_pref_add_choice(ppref, "mimeTeX", "mimetex");
    purple_plugin_pref_add_choice(ppref, "mathTeX", "mathtex");
    purple_plugin_pref_frame_add(frame, ppref);
#endif

    // Print expression
    ppref = purple_plugin_pref_new_with_name_and_label(
            PREFS_PRINTEXPR,
            _("Print the expression along with the image"));
    purple_plugin_pref_frame_add(frame, ppref);

#ifndef _WIN32
    purple_plugin_pref_frame_add(frame,purple_plugin_pref_new_with_label(
                _("Only mimeTeX")));
#endif

    // Reverse image
    ppref = purple_plugin_pref_new_with_name_and_label(
            PREFS_NEGATE,
            _("Negate image. Select for dark background."));
    purple_plugin_pref_frame_add(frame, ppref);

    // Smash expression
    ppref = purple_plugin_pref_new_with_name_and_label(
            PREFS_SMASH,
            _("Smash expression. Horizontally compact."));
    purple_plugin_pref_set_type(ppref, PURPLE_PLUGIN_PREF_CHOICE);
    purple_plugin_pref_add_choice(ppref, _("Default"), "");
    purple_plugin_pref_add_choice(ppref, _("Smash"),    "\\smash");
    purple_plugin_pref_add_choice(ppref, _("No smash"), "\\nosmash");
    purple_plugin_pref_frame_add(frame, ppref);

#ifndef _WIN32
    purple_plugin_pref_frame_add(frame,purple_plugin_pref_new_with_label(
                _("For both renderers")));
#endif

    // Font size
    ppref = purple_plugin_pref_new_with_name_and_label(
            PREFS_FONT_SIZE,
            _("Font size"));
    purple_plugin_pref_set_type(ppref, PURPLE_PLUGIN_PREF_CHOICE);
    purple_plugin_pref_add_choice(ppref, "0", GINT_TO_POINTER(1));
    purple_plugin_pref_add_choice(ppref, "1", GINT_TO_POINTER(1));
    purple_plugin_pref_add_choice(ppref, "2", GINT_TO_POINTER(2));
    purple_plugin_pref_add_choice(ppref, "3", GINT_TO_POINTER(3));
    purple_plugin_pref_add_choice(ppref, "4", GINT_TO_POINTER(4));
    purple_plugin_pref_add_choice(ppref, "5", GINT_TO_POINTER(5));
    purple_plugin_pref_add_choice(ppref, "6", GINT_TO_POINTER(6));
    purple_plugin_pref_add_choice(ppref, "7", GINT_TO_POINTER(7));
    purple_plugin_pref_frame_add(frame, ppref);

    // Style
    ppref = purple_plugin_pref_new_with_name_and_label(
            PREFS_STYLE,
            _("Force style"));
    purple_plugin_pref_set_type(ppref, PURPLE_PLUGIN_PREF_CHOICE);
    purple_plugin_pref_add_choice(ppref, "Auto", "");
    purple_plugin_pref_add_choice(ppref, "Displaystyle", "\\displaystyle");
    purple_plugin_pref_add_choice(ppref, "Textstyle",    "\\textstyle");
    purple_plugin_pref_frame_add(frame, ppref);

    // Font Color
    ppref = purple_plugin_pref_new_with_name_and_label(
            PREFS_FONT_COLOR,
            _("Font color, use default if you have an old version of mimeTeX."));
    purple_plugin_pref_set_type(ppref, PURPLE_PLUGIN_PREF_CHOICE);
    purple_plugin_pref_add_choice(ppref, _("Default"), "");
    purple_plugin_pref_add_choice(ppref, _("Black"),   "black");
    purple_plugin_pref_add_choice(ppref, _("White"),   "white");
    purple_plugin_pref_add_choice(ppref, _("Red"),     "red");
    purple_plugin_pref_add_choice(ppref, _("Cyan"),    "cyan");
    purple_plugin_pref_add_choice(ppref, _("Blue"),    "blue");
    purple_plugin_pref_add_choice(ppref, _("Yellow"),  "yellow");
    purple_plugin_pref_add_choice(ppref, _("Green"),   "green");
    purple_plugin_pref_add_choice(ppref, _("Magenta"), "magenta");
    purple_plugin_pref_frame_add(frame, ppref);

    // Prepend
    ppref = purple_plugin_pref_new_with_name_and_label(
            PREFS_PREPEND,
            _("Prepend all expressions: (example \\gammacorrection{3.5})"));
    purple_plugin_pref_frame_add(frame, ppref);

    return frame;
}

static PurplePluginUiInfo prefs_info = {
    get_plugin_pref_frame,
    0,
    NULL,
    /* Padding */
    NULL,
    NULL,
    NULL,
    NULL
};

static PurplePluginInfo info =
{
    PURPLE_PLUGIN_MAGIC,
    PURPLE_MAJOR_VERSION,
    PURPLE_MINOR_VERSION,
    PURPLE_PLUGIN_STANDARD,                           /**< type           */
    NULL,                                             /**< ui_requirement */
    0,                                                /**< flags          */
    NULL,                                             /**< dependencies   */
    PURPLE_PRIORITY_DEFAULT,                          /**< priority       */
    "core-micket-" PLUGIN_NAME,                       /**< id             */
    PLUGIN_NAME,                                      /**< name           */
    PLUGIN_VERSION,                                   /**< version        */
    /**<  summary        */
    // I'm not allowed to translate this?
    ("To display LaTeX formula into Pidgin conversation."),
    /**<  description    */
    ("Put LaTeX-code between $$ ... $$ markup to have it displayed as a "
        "picture in your conversation.\n"
        "Remember that your contact needs an similar plugin or else he will "
        "just see the pure LaTeX-code.\n"
        "You must have mimeTeX or mathTeX installed (in your PATH)"),
    /**< author */
    "Mikael Öhman <micketeer@gmail.com>\n"
        "Based on LaTeX-plugin by:\n"
        "Benjamin Moll <qjuh@users.sourceforge.net>\n"
        "Nicolas Schoonbroodt <nicolas@ffsa.be>\n"
        "Nicolai Stange <nic-stange@t-online.de>\n",
    "http://pidgintex.googlecode.com",                /**< homepage       */
    plugin_load,                                      /**< load           */
    plugin_unload,                                    /**< unload         */
    NULL,                                             /**< destroy        */
    NULL,                                             /**< ui_info        */
    "http://www.forkosh.com/mimetex.html\n"
    "http://www.forkosh.com/mathtex.html",            /**< extra_info     */
    &prefs_info,
    NULL,
    /* padding */
    NULL,
    NULL,
    NULL,
    NULL
};

static void init_plugin(PurplePlugin *plugin)
{
#ifdef ENABLE_NLS
	bindtextdomain(PLUGIN_NAME, LOCALEDIR);
	bind_textdomain_codeset(PLUGIN_NAME, "UTF-8");
#endif
    purple_prefs_add_none  (PREFS_BASE);
    purple_prefs_add_bool  (PREFS_SENDIMAGE,  FALSE);
    purple_prefs_add_bool  (PREFS_PRINTEXPR,  TRUE);
    purple_prefs_add_bool  (PREFS_NEGATE,     FALSE);
    purple_prefs_add_int   (PREFS_FONT_SIZE,  3);
    purple_prefs_add_string(PREFS_SMASH,      "");
    purple_prefs_add_string(PREFS_STYLE,      "");
    purple_prefs_add_string(PREFS_FONT_COLOR, "");
#ifndef _WIN32
    purple_prefs_add_string(PREFS_RENDERER,   "mimetex");
#endif
    purple_prefs_add_string(PREFS_PREPEND,    "");
}

PURPLE_INIT_PLUGIN(pidginTeX, init_plugin, info)

